### PR TITLE
docker: whitelist entrypoints

### LINKED ISSUE #715: New container entry point could be overriding cml pr command with shell homonym
The new smart entry point introduced with #703 could be overriding `cml pr` the [`pr(1p)`](https://man7.org/linux/man-pages/man1/pr.1p.html) system executable. As per https://github.com/iterative/cml/pull/703#issuecomment-918056150, we should probably consider aliasing only `sh` https://github.com/iterative/cml/pull/703/commits/fb82963f50064ea46044c924d4a57a8971ef1011 or using `CMD` instead.

https://github.com/iterative/cml/blob/4f7a5d95fb75f9dbf052bc9789ca235716560710/Dockerfile#L108

### PR DESCRIPTION
**Pull request by casperdcl**
- e.g. `pr` (fixes #715)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -105,5 +105,5 @@ ENV IN_DOCKER=1
 # also works for GitLab CI/CD
 # https://gitlab.com/gitlab-org/gitlab-runner/-/blob/4c42e96/shells/bash.go#L18-37
 # https://gitlab.com/gitlab-org/gitlab-runner/-/blob/4c42e96/shells/bash.go#L288
-ENTRYPOINT ["/bin/bash", "-c", "which -- \"$0\" &>/dev/null && exec \"$0\" \"$@\" || exec cml \"$0\" \"$@\""]
+ENTRYPOINT ["/bin/bash", "-c", "echo \"$0\" | grep -qE '^(pr|publish|runner|send-(comment|github-check)|tensorboard-dev)$' && exec \"$0\" \"$@\" || exec cml \"$0\" \"$@\""]
```
Comment (0x2b3bfa0) [inline Dockerfile:108]:
```suggestion
ENTRYPOINT ["/bin/bash", "-c", "echo \"$0\" | grep -qE '^(pr|publish|runner|send-(comment|github-check)|tensorboard-dev)$' && exec cml \"$0\" \"$@\" || exec \"$0\" \"$@\""]
```

Either that or `grep -v`, right? 🤔  We want to call `cml` if the command matches, and not the other way around.

Comment (casperdcl) [reply Dockerfile]: doh

File: Dockerfile [hunk 2]
```diff
@@ -105,5 +105,5 @@ ENV IN_DOCKER=1
 # also works for GitLab CI/CD
 # https://gitlab.com/gitlab-org/gitlab-runner/-/blob/4c42e96/shells/bash.go#L18-37
 # https://gitlab.com/gitlab-org/gitlab-runner/-/blob/4c42e96/shells/bash.go#L288
-ENTRYPOINT ["/bin/bash", "-c", "which -- \"$0\" &>/dev/null && exec \"$0\" \"$@\" || exec cml \"$0\" \"$@\""]
+ENTRYPOINT ["/bin/bash", "-c", "echo \"$0\" | grep -qE '^(pr|publish|runner|send-(comment|github-check)|tensorboard-dev)$' && exec cml \"$0\" \"$@\" || exec \"$0\" \"$@\""]
```
Comment (0x2b3bfa0) [inline Dockerfile:108]:
```suggestion
ENTRYPOINT ["/bin/bash", "-c", "echo \"$0\" | grep -qE '^(pr|publish|runner|send-(comment|github-check)|tensorboard-dev|--?(-|\w)+(=.+)?)$' && exec cml \"$0\" \"$@\" || exec \"$0\" \"$@\""]
```

Perhaps also when `$0` is an option?

Comment (casperdcl) [reply Dockerfile:108]: nice idea

```suggestion
ENTRYPOINT ["/bin/bash", "-c", "echo \"$0\" | grep -qE '^(pr|publish|runner|send-(comment|github-check)|tensorboard-dev|--?[[:alnum:]].*)$' && exec cml \"$0\" \"$@\" || exec \"$0\" \"$@\""]
```

Comment (0x2b3bfa0) [reply Dockerfile]: Please note that [`[:alnum:]`](https://www.gnu.org/software/grep/manual/html_node/Character-Classes-and-Bracket-Expressions.html#index-alnum-character-class) doesn't include hyphens like in `--long-option`

Comment (0x2b3bfa0) [reply Dockerfile]: Also, [`\w`](https://www.gnu.org/software/grep/manual/html_node/The-Backslash-Character-and-Special-Expressions.html) should be an alias for `[_[:alnum:]]` and still don't include hyphens.

Comment (casperdcl) [reply Dockerfile]: techincally `--?[[_:alpha:]].*` is most correct but `--?\w.*` is neater.

Comment (0x2b3bfa0) [reply Dockerfile]: It still has some flaws, but it's definitely neater. Let's let `yargs` do teh work for us and print a nice error.

`echo "--_验证错误" | grep -E '^--?\w.*$'`,

**Review by 0x2b3bfa0 — APPROVED** [review]

File: Dockerfile (hunk 2 above)
Comment (casperdcl) [reply Dockerfile]: yar :black_flag:

**Pull request merged**